File: lambda/src/api/upload_photo.py
import base64
import json
import os
import uuid
from datetime import datetime

import boto3
from botocore.exceptions import ClientError

s3_client = boto3.client("s3")
dynamodb = boto3.resource("dynamodb")


def get_current_team_level(team_id):
    """Get the current level for a team."""
    try:
        table = dynamodb.Table(os.environ["DUCK_HUNT_TABLE_NAME"])
        print(f"Querying for team_id: {team_id}")

        # Query using the correct schema
        response = table.query(
            KeyConditionExpression="PK = :pk AND begins_with(SK, :sk)",
            FilterExpression="attribute_not_exists(completed_at)",
            ExpressionAttributeValues={
                ":pk": f"TEAM#{team_id}",
                ":sk": "LEVEL#",
            },
        )

        print(f"Query response: {json.dumps(response, default=str)}")

        if not response.get("Items"):
            print(f"No team level found for team {team_id}")
            return None

        # Find the level with the lowest index
        current_level = min(response["Items"], key=lambda x: x.get("index", 0))
        return current_level

    except Exception as e:
        print(f"Error fetching current level for team {team_id}: {str(e)}")
        raise Exception(f"Failed to get current level for team: {team_id}")


def lambda_handler(event, context):
    try:
        # Extract headers
        headers = event.get("headers", {})
        user_id = headers.get("user-id")
        team_id = headers.get("team-id")

        # Validate required headers
        if not all([user_id, team_id]):
            return {
                "statusCode": 400,
                "body": json.dumps(
                    {"error": "Missing required headers (user-id or team-id)"}
                ),
            }

        # Get current level for team
        try:
            current_level = get_current_team_level(team_id)
            if not current_level:
                return {
                    "statusCode": 404,
                    "body": json.dumps(
                        {"error": "No active level found for team"}
                    ),
                }
            level_id = current_level.get("level_id")
        except Exception as e:
            print(f"Error getting current level: {str(e)}")
            return {
                "statusCode": 500,
                "body": json.dumps({"error": "Failed to get current level"}),
            }

        # Check if body exists and is base64 encoded
        if not event.get("body") or not event.get("isBase64Encoded"):
            return {
                "statusCode": 400,
                "body": json.dumps({"error": "Invalid request body"}),
            }

        # Decode base64 body
        body = base64.b64decode(event["body"])

        # Generate unique identifiers and timestamps
        photo_id = str(uuid.uuid4())
        current_time = datetime.now().isoformat()
        epoch_timestamp = int(datetime.now().timestamp())
        sort_key = f"PHOTO#{epoch_timestamp}#{photo_id}"

        # Generate S3 filename
        filename = f"{team_id}/{level_id}/{epoch_timestamp}_{photo_id}.jpg"

        # Upload to S3
        try:
            s3_client.put_object(
                Bucket=os.environ["PHOTO_BUCKET"],
                Key=filename,
                Body=body,
                ContentType="image/jpeg",
            )
        except ClientError as e:
            print(f"Error uploading to S3: {str(e)}")
            return {
                "statusCode": 500,
                "body": json.dumps(
                    {"error": "Failed to upload photo to storage"}
                ),
            }

        # Create DynamoDB item
        try:
            table = dynamodb.Table(os.environ["DUCK_HUNT_TABLE_NAME"])

            item = {
                "PK": f"USER#{user_id}",
                "SK": sort_key,
                "GSI1PK": f"TEAM#{team_id}",
                "GSI1SK": sort_key,
                "GSI3PK": f"LEVEL#{level_id}",
                "GSI3SK": sort_key,
                "ItemType": "PHOTO",
                "id": photo_id,
                "user_id": user_id,
                "team_id": team_id,
                "level_id": level_id,
                "url": filename,
                "created_at": current_time,
                "updated_at": current_time,
            }

            table.put_item(Item=item)

            # Return photo data
            photo_data = {
                "id": photo_id,
                "user_id": user_id,
                "team_id": team_id,
                "level_id": level_id,
                "url": filename,
                "created_at": current_time,
                "updated_at": current_time,
            }

            return {
                "statusCode": 200,
                "body": json.dumps(
                    {
                        "message": "Photo uploaded successfully",
                        "photo": photo_data,
                    }
                ),
                "headers": {
                    "Content-Type": "application/json",
                    "Access-Control-Allow-Origin": "*",
                    "Access-Control-Allow-Credentials": "true",
                },
            }

        except ClientError as e:
            print(f"Error updating DynamoDB: {str(e)}")
            return {
                "statusCode": 500,
                "body": json.dumps({"error": "Failed to create photo record"}),
            }

    except Exception as e:
        print(f"Unexpected error: {str(e)}")
        return {
            "statusCode": 500,
            "body": json.dumps({"error": "Internal server error"}),
        }


# Test event
# {
#   "headers": {
#     "content-type": "multipart/form-data; boundary=----WebKitFormBoundaryXXXXXX",
#     "user-id": "123e4567-e89b-12d3-a456-426614174000",
#     "team-id": "987fcdeb-51a2-12d3-a456-426614174000"
#   },
#   "body": "/9j/4AAQSkZJRgABAQEAYABgAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkSEw8UHRofHh0aHBwgJC4nICIsIxwcKDcpLDAxNDQ0Hyc5PTgyPC4zNDL/2wBDAQkJCQwLDBgNDRgyIRwhMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjL/wAARCAABAAEDASIAAhEBAxEB/8QAFQABAQAAAAAAAAAAAAAAAAAAAAn/xAAUEAEAAAAAAAAAAAAAAAAAAAAA/8QAFQEBAQAAAAAAAAAAAAAAAAAAAAX/xAAUEQEAAAAAAAAAAAAAAAAAAAAA/9oADAMBAAIRAxEAPwCdABmX/9k=",
#   "isBase64Encoded": true
# }

File: lambda/src/api/uploadPhoto.ts
import { PutObjectCommand, S3Client } from "@aws-sdk/client-s3";
import { corsHeaders, RequestHeaders } from "@shared/types";
import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { PhotoOperations } from "src/dynamo/photo";
import { v4 as uuidv4 } from "uuid";
import { fetchBaseData } from "./fetchBaseData";

const s3Client = new S3Client({ region: process.env.AWS_REGION });

interface UploadPhotoRequest {
  photo: string; // base64 encoded image data
  filename: string;
  contentType: string;
  size: number;
}

const detectImageFormat = (
  base64Data: string,
  contentType?: string
): { extension: string; mimeType: string } => {
  // First try to use the provided content type
  if (contentType) {
    if (contentType.includes("jpeg") || contentType.includes("jpg")) {
      return { extension: "jpg", mimeType: "image/jpeg" };
    }
    if (contentType.includes("png")) {
      return { extension: "png", mimeType: "image/png" };
    }
    if (contentType.includes("gif")) {
      return { extension: "gif", mimeType: "image/gif" };
    }
  }

  // Fallback to checking the base64 data
  const buffer = Buffer.from(base64Data, "base64");

  // Check for JPEG
  if (
    buffer.length >= 3 &&
    buffer[0] === 0xff &&
    buffer[1] === 0xd8 &&
    buffer[2] === 0xff
  ) {
    return { extension: "jpg", mimeType: "image/jpeg" };
  }

  // Check for PNG
  if (
    buffer.length >= 8 &&
    buffer[0] === 0x89 &&
    buffer[1] === 0x50 &&
    buffer[2] === 0x4e &&
    buffer[3] === 0x47 &&
    buffer[4] === 0x0d &&
    buffer[5] === 0x0a &&
    buffer[6] === 0x1a &&
    buffer[7] === 0x0a
  ) {
    return { extension: "png", mimeType: "image/png" };
  }

  // Default to JPEG
  return { extension: "jpg", mimeType: "image/jpeg" };
};

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  console.log("INFO: Received event:", JSON.stringify(event, null, 2));

  try {
    const headers = event.headers as unknown as RequestHeaders;

    const { currentLevel, gameId, currentTeamLevel, userMessages } =
      await fetchBaseData(headers);

    // Parse request body
    if (!event.body) {
      console.error("ERROR: Request body is missing or empty");
      throw new Error("Request body is required");
    }

    const requestBody: UploadPhotoRequest = JSON.parse(event.body);
    console.log("INFO: Parsed request body:", {
      filename: requestBody.filename,
      contentType: requestBody.contentType,
      size: requestBody.size,
      photoDataLength: requestBody.photo?.length || 0,
    });

    // Validate required fields
    if (!requestBody.photo) {
      throw new Error("Photo data is required");
    }

    // Convert base64 to buffer
    const fileBuffer = Buffer.from(requestBody.photo, "base64");
    console.log(
      `INFO: Converted base64 to buffer, size: ${fileBuffer.length} bytes`
    );

    // Detect image format
    const { extension, mimeType } = detectImageFormat(
      requestBody.photo,
      requestBody.contentType
    );
    console.log(`INFO: Detected image format: ${extension} (${mimeType})`);

    // Generate unique identifiers and timestamps
    console.log("INFO: Generating unique identifiers");
    const photoId = uuidv4();
    const epochTimestamp = Math.floor(Date.now() / 1000);

    // Generate S3 filename
    const filename = `${currentTeamLevel.team_id}/${currentTeamLevel.level_id}/${epochTimestamp}_${photoId}.${extension}`;
    console.log(`INFO: Generated S3 filename: ${filename}`);

    // Check environment variables
    const photoBucket = process.env.PHOTO_BUCKET;
    if (!photoBucket) {
      console.error("ERROR: PHOTO_BUCKET environment variable not set");
      throw new Error("PHOTO_BUCKET environment variable is required");
    }

    console.log(`INFO: Using S3 bucket: ${photoBucket}`);

    // Upload to S3
    try {
      console.log("INFO: Starting S3 upload...");
      const putCommand = new PutObjectCommand({
        Bucket: photoBucket,
        Key: filename,
        Body: fileBuffer,
        ContentType: mimeType,
      });

      await s3Client.send(putCommand);
      console.log("INFO: S3 upload successful");
    } catch (error) {
      console.error("ERROR: S3 upload failed:", error);
      throw new Error(
        `Failed to upload photo to S3: ${
          error instanceof Error ? error.message : "Unknown error"
        }`
      );
    }

    // Save photo metadata to DynamoDB
    console.log("INFO: Saving photo metadata to DynamoDB...");
    await PhotoOperations.create({
      game_id: gameId,
      team_id: currentTeamLevel.team_id,
      level_id: currentTeamLevel.level_id,
      user_id: headers["user-id"],
      url: `https://${photoBucket}.s3.amazonaws.com/${filename}`,
    });

    console.log("INFO: Photo metadata saved to DynamoDB");

    return {
      statusCode: 200,
      headers: corsHeaders,
      body: JSON.stringify({
        success: true,
        message: "Photo uploaded successfully",
        photo: {
          id: photoId,
          url: `https://${photoBucket}.s3.amazonaws.com/${filename}`,
          format: extension,
        },
      }),
    };
  } catch (error) {
    console.error("ERROR: Unexpected top-level error:", error);
    return {
      statusCode: 500,
      headers: corsHeaders,
      body: JSON.stringify({
        success: false,
        error: error instanceof Error ? error.message : "Internal server error",
      }),
    };
  }
};
